modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb

```diff
@@ -61,7 +61,11 @@ def legacy_classes_format(response)
 
   def get_classes(environment)
     future = async_get_classes(environment)
-    future.value!(MAX_CLIENT_TIMEOUT)
+    cache_used = @m.synchronize { !!@etag_cache[environment] } #etags are only available when classes cache is enabled
+    client_timeout = cache_used ? DEFAULT_CLIENT_TIMEOUT : 2 * DEFAULT_CLIENT_TIMEOUT
```

Comment:
> so I don't think the default behaviour of the smart proxy can be too aggressive in assuming it is.

Addendum... the default behaviour perhaps isn't so important, but it should be possible for the user to configure the smart proxy to work well when caching is disabled. It's very much valid to use Puppet Server without caching. For class imports, both Foreman's rake task and API mean the Foreman UI responsiveness isn't very important, it's normal for them to block.

If the smart proxy's timeout can be increased through a config to work well in very large environments then a lower default seems reasonable. Users with large environments can then choose between reconfiguring the timeout or enabling caching.